doc/manual_experimental.rst

```diff
@@ -418,1296 +427,1139 @@ here.
+      MyObject = object
+        enumerables: seq[IntEnumerable]
+        streams: seq[OutputStream.vtref]
 
-The following example shows how some form of hoisting can be implemented:
+    proc addEnumerable(o: var MyObject, e: IntEnumerable) =
+      o.enumerables.add e
 
-.. code-block:: nim
+    proc addStream(o: var MyObject, e: OutputStream.vtref) =
+      o.streams.add e
 
-  import std/pegs
+  The procs that will be included in the vtable are derived from the concept
+  body and include all proc calls for which all param types were specified as
+  concrete types. All such calls should include exactly one param of the type
+  matched against the concept (not necessarily in the first position), which
+  will be considered the value bound to the vtable.
 
-  template optPeg{peg(pattern)}(pattern: string{lit}): Peg =
-    var gl {.global, gensym.} = peg(pattern)
-    gl
+  Overloads will be created for all captured procs, accepting the vtable type
+  in the position of the captured underlying object.
 
-  for i in 0 .. 3:
-    echo match("(a b c)", peg"'(' @ ')'")
-    echo match("W_HI_Le", peg"\y 'while'")
+  Under these rules, it's possible to obtain a vtable type for a concept with
+  unbound type parameters or one instantiated with metatypes (type classes),
+  but it will include a smaller number of captured procs. A completely empty
+  vtable will be reported as an error.
 
-The `optPeg` template optimizes the case of a peg constructor with a string
-literal, so that the pattern will only be parsed once at program startup and
-stored in a global `gl` which is then re-used. This optimization is called
-hoisting because it is comparable to classical loop hoisting.
+  The `vtref` magic produces types which can be bound to `ref` types and
+  the `vtptr` magic produced types bound to `ptr` types.
 
 
-AST based overloading
+Type bound operations
 =====================
 
-Parameter constraints can also be used for ordinary routine parameters; these
-constraints affect ordinary overloading resolution then:
-
-.. code-block:: nim
-
-  proc optLit(a: string{lit|`const`}) =
-    echo "string literal"
-  proc optLit(a: string) =
-    echo "no string literal"
-
-  const
-    constant = "abc"
+There are 4 operations that are bound to a type:
 
-  var
-    variable = "xyz"
+1. Assignment
+2. Moves
+3. Destruction
+4. Deep copying for communication between threads
 
-  optLit("literal")
-  optLit(constant)
-  optLit(variable)
+These operations can be *overridden* instead of *overloaded*. This means the
+implementation is automatically lifted to structured types. For instance if type
+`T` has an overridden assignment operator `=` this operator is also used
+for assignments of the type `seq[T]`. Since these operations are bound to a
+type they have to be bound to a nominal type for reasons of simplicity of
+implementation: This means an overridden `deepCopy` for `ref T` is really
+bound to `T` and not to `ref T`. This also means that one cannot override
```

Comment:
```suggestion
These operations can be *overridden* instead of *overloaded*. This means that the
implementation is automatically lifted to structured types. For instance, if type
`T` has an overridden assignment operator `=`, this operator is also used
for assignments of the type `seq[T]`. Since these operations are bound to a
type, they have to be bound to a nominal type for reasons of simplicity of
implementation: This means an overridden `deepCopy` for `ref T` is really
bound to `T` and not to `ref T`. This also means that one cannot override
```